// connect-architect/connect/connect-plugin.js
var utils = require("./connect-utils-shim");
var netutil = require("netutil");
var connect = require("connect");
var cookieParser = require("cookie-parser");
var bodyParser = require("body-parser");
var http = require("http");
var https = require("https");

// connect@3 removed every bundled middleware and the legacy `createServer`
// alias that getModule() consumers across the repo still reach (compress,
// favicon, basicAuth, static, createServer -- plus the unwired cookieParser/
// urlencoded/json/query). Expose a compatibility shim from getModule() that
// maps each legacy name onto its extracted package so NO consumer file needs
// editing. Heavy replacements are require()d lazily (inside getters / factory
// bodies) so this load-bearing module still LOADS even if an optional sibling
// package is briefly absent; the require only fires -- exactly where the old
// connect.<name> would have -- when a consumer actually calls the name.
var connectModule = function() { return connect.apply(this, arguments); };
connectModule.createServer = connect;
Object.defineProperty(connectModule, "static", {
    configurable: true,
    get: function() { return require("serve-static"); }
});
Object.defineProperty(connectModule, "compress", {
    configurable: true,
    get: function() { return require("compression"); }
});
Object.defineProperty(connectModule, "favicon", {
    configurable: true,
    get: function() { return require("serve-favicon"); }
});
// Unwired-but-cheap insurance (no new deps) for the connect.cookieparser/
// connect.bodyparser/connect.query architect plugins if a non-standalone/cdn
// config ever loads them. logger/session are intentionally NOT mapped (would
// need morgan/express-session) and are confirmed unwired.
connectModule.cookieParser = cookieParser;
connectModule.urlencoded = function() { return bodyParser.urlencoded({ extended: true, limit: "1mb" }); };
connectModule.json = function() { return bodyParser.json({ limit: "1mb" }); };
connectModule.query = function() {
    var qs = require("qs");
    return function(req, res, next) {
        if (!req.query) {
            var i = req.url.indexOf("?");
            req.query = i >= 0 ? qs.parse(req.url.slice(i + 1)) : {};
        }
        next();
    };
};
// basicAuth: constant-time replacement for connect@2 basicAuth's `==` compare.
// Hashing both sides to a fixed 32-byte sha256 digest sidesteps
// timingSafeEqual's equal-length requirement (wrong-length creds compare
// safely and never throw). Same (user, pass) -> (req,res,next) call contract.
connectModule.basicAuth = function(username, password) {
    var auth = require("basic-auth");
    var crypto = require("crypto");
    function safeCompare(input, expected) {
        return crypto.timingSafeEqual(
            crypto.createHash("sha256").update(String(input)).digest(),
            crypto.createHash("sha256").update(String(expected)).digest()
        );
    }
    return function(req, res, next) {
        var credentials = auth(req);
        var ok = false;
        if (credentials) {
            var userOk = safeCompare(credentials.name, username);
            var passOk = safeCompare(credentials.pass, password);
            ok = userOk && passOk;
        }
        if (ok) return next();
        res.statusCode = 401;
        res.setHeader("WWW-Authenticate", 'Basic realm="Authorization Required"');
        res.end("Unauthorized");
    };
};

module.exports = function startup(options, imports, register) {
    var globalOptions = options.globals ? merge([options.globals]) : {};
    
    var requestMethods = {};
    var responseMethods = {};
    
    var app = connect();

    var hookNames = [
        "Start",
        "Setup",
        "Main",
        "Session"
    ];
    var api = {
        getModule: function() {
            return connectModule;
        },
        getUtils: function() {
            return utils;
        },
        /**
         * set per request options. Used e.g. for the view rendering
         */ 
        setOptions: function(options) {
            return function(req, res, next) {
                res.setOptions(options);
                next();
            };
        },
        resetOptions: function() {
            return function(req, res, next) {
                res.resetOptions();
                next();
            };
        },
        setGlobalOption: function(key, value) {
            globalOptions[key] = value;
        },
        getGlobalOption: function(key) {
            return globalOptions[key];
        },
        addResponseMethod: function(name, method) {
            responseMethods[name] = method;
        },
        addRequestMethod: function(name, method) {
            requestMethods[name] = method;
        }

    };
    
    api.addResponseMethod("setOptions", function(options) {
        this._options = this._options || [];
        this._options.push(options);
    });
    
    api.addResponseMethod("getOptions", function(options) {
        var opts = [globalOptions].concat(this._options || []);
        if (options)
            opts = opts.concat(options);
        
        return merge(opts);
    });
    
    api.addResponseMethod("resetOptions", function() {
        if (this._options)
            this._options.pop();
    });

    
    hookNames.forEach(function(name) {
        var hookServer = connect();
        app.use(hookServer);
        api["use" + name] = hookServer.use.bind(hookServer);
    });

    var connectHook = connectError();
    app.use(connectHook);
    api.useError = connectHook.use;

    api.useSetup(cookieParser());
    api.useSetup(bodyParser.urlencoded({ extended: true, limit: "1mb" }));
    api.useSetup(bodyParser.json({ limit: "1mb" }));

    api.addRoute = app.addRoute;
    api.use = api.useMain;

    api.on = app.on;
    api.emit = app.emit;

    api.useStart(function(req, res, next) {
        for (var name in requestMethods)
            req[name] = requestMethods[name];
            
        for (var name in responseMethods)
            res[name] = responseMethods[name];
            
        next();
    });

    function startListening (port, host) {
        api.getPort = function () {
            return port;
        };
        api.getHost = function () {
            return host;
        };

        var server;
        var proto;
        if (options.secure) {
            proto = "https";
            server = https.createServer(options.secure , app);
        }
        else {
            proto = "http";
            server = http.createServer(app);
        }

        server.listen(port, host, function(err) {
            if (err)
                return register(err);

            console.log("Connect server listening at " + proto + "://"
                + (host == "0.0.0.0" && options.showRealIP
                ? getLocalIPs()[0]
                : host) + ":" + port);

            register(null, {
                "onDestruct": function(callback) {
                    app.close();
                    app.on("close", callback);
                },
                "connect": api,
                "http": {
                    getServer: function() {
                        return server;
                    }
                }
            });
        });
        
        if (options.websocket)
            attachWsServer(server, app);
        
        function attachWsServer(server, app) {
            server.on("upgrade", function(req, socket, head) {
                var res = new http.ServerResponse(req);
                req.ws = {
                    socket: socket,
                    head: head
                };
                req.method = "UPGRADE";
                
                res.write = function() {
                    // console.log("RES WRITE", arguments);
                };
                res.writeHead = function() {
                    // console.log("RES WRITEHEAD", arguments);
                };
                res.end = function() {
                    // console.log("RES END", arguments);
                    socket.end();
                };
                
                app.handle(req, res, function(err) {
                    if (err) {
                        console.error("Websocket error", err);
                    }
                    
                    socket.end();
                });
            });
        }
    }

    if (options.port instanceof Array) {
        netutil.findFreePort(options.port[0], options.port[1], options.host, function(err, port) {
            if (err)
                return register(err);

            startListening(port, options.host || "localhost");
        });
    } else {
        startListening(options.port, options.host || "localhost");
    }
    
    function connectError() {
        var filters = [];
    
        function handle(err, req, res, out) {
            var rest = filters.concat();
    
            function next(err) {
                var filter = rest.shift();
                if (!filter)
                    return out(err);
    
                filter(err, req, res, next);
            }
            next(err);
        }
    
        handle.use = function(middleware) {
            filters.push(middleware);
        };
        
        return handle;
    }
    
    function getLocalIPs() {
        var os = require("os");
    
        var interfaces = os.networkInterfaces ? os.networkInterfaces() : {};
        var addresses = [];
        for (var k in interfaces) {
            for (var k2 in interfaces[k]) {
                var address = interfaces[k][k2];
                if (address.family === "IPv4" && !address.internal) {
                    addresses.push(address.address);
                }
            }
        }
        return addresses;
    }
};

function merge(objects) {
    var result = {};
    for (var i=0; i<objects.length; i++) {
        var obj = objects[i];
        for (var key in obj) {
            result[key] = obj[key];
        }
    }
    return result;
}
