// connect-architect/connect/connect-utils-shim.js
/*
 * Minimal compatibility shim for the handful of helpers that the Cloud9
 * connect plugin (exposed via getUtils()) and the legacy vendored static
 * middleware referenced from connect@2's "connect/lib/utils", which connect@3
 * removed. Self-contained (no external deps); implementations mirror connect 2.x
 * but read response headers via the modern res.getHeader API.
 */
"use strict";

exports.error = function(code, msg) {
    var err = new Error(msg || require("http").STATUS_CODES[code]);
    err.status = code;
    return err;
};

exports.escape = function(html) {
    return String(html)
        .replace(/&(?!\w+;)/g, "&amp;")
        .replace(/</g, "&lt;")
        .replace(/>/g, "&gt;")
        .replace(/"/g, "&quot;");
};

exports.badRequest = function(res) {
    res.statusCode = 400;
    res.end("Bad Request");
};

exports.forbidden = function(res) {
    var body = "Forbidden";
    res.statusCode = 403;
    res.setHeader("Content-Type", "text/plain");
    res.setHeader("Content-Length", body.length);
    res.end(body);
};

exports.removeContentHeaders = function(res) {
    var headers = (typeof res.getHeaderNames === "function")
        ? res.getHeaderNames()
        : Object.keys(res._headers || {});
    headers.forEach(function(field) {
        if (field.toLowerCase().indexOf("content") === 0)
            res.removeHeader(field);
    });
};

exports.conditionalGET = function(req) {
    return req.headers["if-modified-since"] || req.headers["if-none-match"];
};

exports.notModified = function(res) {
    exports.removeContentHeaders(res);
    res.statusCode = 304;
    res.end();
};

exports.etag = function(stat) {
    return '"' + stat.size + "-" + Number(stat.mtime) + '"';
};

exports.modified = function(req, res, headers) {
    headers = headers || {};
    var get = typeof res.getHeader === "function"
        ? function(n) { return res.getHeader(n); }
        : function() { return undefined; };
    var modifiedSince = req.headers["if-modified-since"];
    var lastModified = headers["last-modified"] || get("Last-Modified");
    var noneMatch = req.headers["if-none-match"];
    var etag = headers["etag"] || get("ETag");
    if (noneMatch) noneMatch = noneMatch.split(/ *, */);
    if (noneMatch && etag && ~noneMatch.indexOf(etag)) return false;
    if (modifiedSince && lastModified) {
        modifiedSince = new Date(modifiedSince);
        lastModified = new Date(lastModified);
        if (!isNaN(modifiedSince.getTime()) && lastModified <= modifiedSince)
            return false;
    }
    return true;
};

exports.parseRange = function(size, str) {
    var valid = true;
    str = String(str).substr(6);
    var arr = str.split(",").map(function(range) {
        range = range.split("-");
        var start = parseInt(range[0], 10);
        var end = parseInt(range[1], 10);
        if (isNaN(start)) {
            start = size - end;
            end = size - 1;
        } else if (isNaN(end)) {
            end = size - 1;
        }
        if (isNaN(start) || isNaN(end) || start > end || start < 0)
            valid = false;
        return { start: start, end: end };
    });
    return valid ? arr : undefined;
};
